identity/aziot-identityd/src/identity.rs

```diff
@@ -893,6 +893,14 @@ impl IdentityManager {
                         std::collections::BTreeSet::default()
                     };
 
+                // Write out device state and settings.
+                // This overwrites any existing device state and settings backup.
+                std::fs::write(prev_device_info_path, device_status)
+                    .map_err(|err| Error::Internal(InternalError::SaveDeviceInfo(err)))?;
+
+                std::fs::write(prev_settings_path, &settings_serialized)
```

Comment:
If creating module identities after saving the backup fails, identityd restarts, won't it think module identities already exist in cloud because the backup matches the settings, and thus it'll skip trying to create them again? I assume that's why the cloud stuff was done before saving the backup.

So rather than move this up, you could just return the early return altogether. In fact, currently it allows cloud to have extra identities that aren't present in the config or in the backup; removing the early return would allow cleaning those up too.

Reply:
Fixed.

I remember discussing this with Rajeev and he mentioned there might be a scenario where users create identities in IoT Hub not managed by Identity Service, and we don't want IS to delete those. So that's why there's an early return there. Does that make sense?